### DISCUSSION TITLE
클로저를 간단한 코드 예시를 들어 설명해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#55] [스코프]
- 클로저를 간단한 예시를 들어 설명해주세요.

**Comment by pipisebastian** [comment]
```tsx
const calculatorFn = () => {
  let count = 0;
  const increase = () => {
    return ++count;
  }
  const decrease = () => {
    return --count;
  }
  return {increase, decrease}
}

const calculator = calculatorFn();
console.log(calculator.increase());
console.log(calculator.increase());
console.log(calculator.decrease());
```

- 외부함수(calculatorFn)의 내부함수(increase, decrease)는 외부함수보다 오래 살아있기 때문에, 외부함수의 렉시컬환경에 접근할 수 있습니다. 이를 클로저라고 하며, 이를 통해 데이터 은닉을 구현할 수 있습니다. count의 경우, increase(), decrease()를 통해서만 접근가능합니다.

**Comment by publdaze** [comment]
> - 클로저를 간단한 예시를 들어 설명해주세요.

```js
const outer = () => {
  const x = 1;
  const inner = () => {
    console.log(x);
  }
  return inner;
}

const func = outer();
func();
```

- 클로저는 함수가 중첩되어 있을 때 반환된 내부함수(중첩함수)가 렉시컬 환경의 스코프를 기억하여 스코프 밖에서 호출되어서 그 스코프에 접근할 수 있는 함수
- 따라서 위의 코드에서 outer의 반환값인 inner 함수가 func에 반환 값으로 저장이 된 후 상위 스코프의 outer 함수 스코프 밖에서 호출되었지만 outer 함수 스코프에 접근이 가능해서 1이 출력될 수 있음
